enhancements/installer/single-node-installation-bootstrap-in-place.md

```diff
@@ -0,0 +1,256 @@
+Post reboot:
+The kubelet service will start the control plane static pods.
+Kubelet will send a CSR (see below) and join the cluster.
+CVO will deploy all cluster operators.
+The control plane operators will rollout a new revision (if necessary).  
+
+#### Openshift-installer
+
+We will add logic to the installer to create `bootstrap-in-place.ign` ignition config.
+This ignition config will diverge from the bootstrap ignitionin:
+bootkube:
+1. Start cluster-bootstrap without required pods (`--required-pods=''`)
+2. Run cluster-bootstrap with iBIP entrypoint to enrich the master ignition.
+3. Write RHCOS image and the master ignition to disk.
+4. Reboot the node.
+
+#### Cluster-bootstrap
+
+By default, `cluster-bootstrap` start the bootstrap control plane and create all the manifests under /opt/openshift/manifests.
+`cluster-bootstrap` also wait for a list required pods to be ready, these pods are expetced to start running on the master nodes.
+In case we are running the bootstrap in place, there is no master node that can run the pods. cluster-bootstrap should apply the manifest and tear down the control plane, in case it fail to apply some of the manifest it should return an error.
+
+
+`cluster-bootstrap` will have a new entrypoint `iBIP` which will get the master ignition as input and will enrich the master ignition with control plane static pods manifests and all required resources including etcd data.
+
+#### Bootstrap / Control plane static pods
+The control plane components we will add to the master ignition are (to be placed under `/etc/kubernetes/manifests`):
+1. etcd-pod
+2. kube-apiserver-pod
+3. kube-controller-manager-pod
+4. kube-scheduler-pod
+
+Control plane required resources to be added to the ignition:
+1. `/var/lib/etcd`
+2. `/etc/kubernetes/bootstrap-configs`
+3. /opt/openshift/tls/* (`/etc/kubernetes/bootstrap-secrets`)
+4. /opt/openshift/auth/kubeconfig-loopback (`/etc/kubernetes/bootstrap-secrets/kubeconfig`)
+
+Note: `/etc/kubernetes/bootstrap-secrets` and `/etc/kubernetes/bootstrap-configs` will be deleted post reboot once the OCP control plane is ready.
+
+The control plane opertors (that will run on the node post reboot) will manage the rollout of new revision of the control plane pods
+
+#### Etcd data
+
+In order to add a viable, working etcd post reboot.
+We will take a snapshot of etcd and add it to the master ignition.
+Oost reboot we will use restore etcd-member from the snapshot
+
+Another option is to stop the etcd pod (move the static pod manifest from `/etc/kubernetes/manifests`).
+Let etcd save it's state and exit, once etcd is down we will add the `/var/lib/etcd` directory to the master ignition.
+After the reboot etcd should start with all the data it had prior to the reboot.
+
+#### Post reboot
+
+We will add a new post-reboot service for approving the kubelet and the node Certificate Sign Requests.
+This service will also cleanup the bootstrap static pods resources once the OCP control plane is ready.
```

Comment:
Is it fair to say then that "installation should result a clean Single Node Openshift, without bootstrap leftovers" is achieved by deleting those bootstrap leftovers?

I feel like there's a fundamental concern about "bootstrap leftovers" that could be discussed in more detail in this doc, and why this proposal is a sufficient mitigation for that concern

Reply:
Lets start with writing down what are the leftovers (if such).

Reply:
all of [the bootstrap services][1]?  The initial YAML manifests?  Bunch of [`/etc` stuff][2]?  Assorted [`*.done` files][3]?  It's possible that you can address some of that by adjusting the bootstrap scripts to explicitly use `/tmp` or something that gets wiped on reboot for scratch space (maybe they already do), but you will likely need something to address the systemd services and `/etc` bits, and possibly other things I'm not remembering at the moment ;).

[1]: https://github.com/openshift/installer/tree/ec982ecf5bd847add0f3af82b04f32251972701c/data/data/bootstrap/systemd/units
[2]: https://github.com/openshift/installer/tree/ec982ecf5bd847add0f3af82b04f32251972701c/data/data/bootstrap/files/etc
[3]: https://github.com/openshift/installer/blob/ec982ecf5bd847add0f3af82b04f32251972701c/data/data/bootstrap/files/usr/local/bin/bootkube.sh.template#L47

Reply:
Since we start with a liveCD none of the `bootstrap services`, `etc stuff` or `/opt/openshift`  "garbage" is left on the node (post reboot).
The files that we need to delete are under:
/etc/kubernetes/bootstrap-secrets` and `/etc/kubernetes/bootstrap-configs`
These files are required for the bootstrap control plane to start before it's replaced by the control plane operators.
Once the OCP control plane static pods are deployed we should cleanup.

Reply:
I think what @eranco74 wrote here would be good as the mitigation for the concern that @markmc raised, and that could all be added to the relevant section around line 167.